### PR TITLE
Weak type detection

### LINKED ISSUE #7485: All Optional Object Interface Converts to any
I couldn't find a relevant issue, but this does seem rather fundamental and against the principle of:

> 1. Statically identify constructs that are likely to be errors.

**TypeScript Version:** 

1.8.7

[**Code**](http://www.typescriptlang.org/Playground#src=interface%20Optional%20%7B%0A%09foo%3F%3A%20string%3B%0A%7D%0A%0Afunction%20foo%28o%3A%20Optional%29%3A%20void%20%7B%7D%0A%0Afoo%28'bar'%29%3B%0Afoo%281%29%3B%0Afoo%28%7B%20bar%3A%20'baz'%20%7D%29%3B%0A)

``` ts
interface Optional {
    foo?: string;
}

function foo(o: Optional): void {}

foo('bar'); // should throw
foo(1); // should throw
foo({ bar: 'baz' }); // should (and does) throw
```

**Expected behavior:**

Expected behaviour is that you should be able to type guard against other non-objects and only accept Objects with no properties or declared properties.

**Actual behavior:** 

Passing anything other than an object literal that contains extra properties is acceptable.

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #7485 

Builds on #3842 by correctly handling intersections that contain weak types but may not necessarily be weak themselves. I ran the change on our real-world code corpus and on DefinitelyTyped and verified that it doesn't cause a huge number of breaks.

Weak types are
1. Object types with at least one property
2. Where all properties are optional
3. And that do not have a string index signature, number index signature, call signature or construct signature.

Weak types are common in Javascript, but the compiler allows almost anything to be assigned to them. The compiler misses a lot of obvious errors when you pass values to weakly typed parameters, implement weakly typed interfaces, or even cast to weak types. Here is an example of a weak type:

```ts
interface Options {
  flag?: boolean
  maximum?: number
  filename?: string
}
```

The weak type check occurs when the target is a weak type and the source has at least one property. The weak type check issues an error if the source does not have any of the properties of the target. This is no longer allowed:

```ts
declare function setup(options: Options): void
const payload = {
  id = 1
  name = "foo.txt"
  data = JSON.stringify(something)
}
setup(payload) // weak type error!
```

Intersection types are handled specially as weak type targets. Individual constituents of an intersection are exempt from the weak type check. However, if all constituents of an intersection are weak, then the weak type check uses the combined properties of the intersection as the target.

```ts
interface UIOptions {
  greener?: boolean
  green?: number
}
type AllOptions = Options & UIOptions
declare function setupUI(options: AllOptions): void
setup(payload) // still not allowed!
```

The global object type is handled specially as a weak type source. It is not an error to assign `Object` to a weak type, even if they don't share any properties. That's because people think of `Object` as similar to `{}`, even though it actually includes a number of properties.

This PR also improves errors for JSX attributes, which previously implemented their own weak type detection. It also improves subtype reduction because a weak type is no longer treated as a universal supertype like `{}` is.

### CODE REVIEW

**Comment by sandersn** [comment]
The Travis failure occurs because of an error in react.d.ts, which our tests have a copy of. Component implements a weak type named ComponentLifecycle but fails to include any of the weak type's methods. I'll send a PR to definitely typed tomorrow to fix the official version and also update this PR.

**Comment by RyanCavanaugh** [comment]
> Component implements a weak type named ComponentLifecycle but fails to include any of the weak type's methods.

Where's the head-desk emoji when you need it

**Comment by sandersn** [comment]
The weird test changes in controlFlowInstanceOf, subTypeReduction* and generatorTypeCheck63 are because subtype reduction has changed. `{ foo?: Foo }` was previously a supertype of `{ bar: Bar }` by virtue of being a weak type, and of course `{}` is the supertype of all object types. With this PR, weak types are only supertypes of object types that share at least one property with them.

So unions produce a different type when asked to do subtype reduction:

```ts
{ foo?: Foo } | { bar: Bar }
//   used to reduce to
{ foo?: Foo }
// now it does not!
```

I don't think it was supposed to work this way according to the spec, so this is an improvement.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8977,16 +8978,35 @@ namespace ts {
...
                 for (const targetType of targetTypes) {
                     const related = isRelatedTo(source, targetType, reportErrors);
                     if (!related) {
+                        dynamicDisableWeakTypeErrors = saveDynamicDisableWeakTypeErrors;
                         return Ternary.False;
                     }
                     result &= related;
                 }
+                dynamicDisableWeakTypeErrors = saveDynamicDisableWeakTypeErrors;
+                if (source !== globalObjectType && getPropertiesOfType(source).length > 0 && every(target.types, isWeak)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Comment on what you're doing here :)

Comment (sandersn) [inline src/compiler/checker.ts]: I refactored this into a separate function and a named boolean. I think the two new names should serve as adequate documentation.

**Comment by sandersn** [comment]
# Summary of Real World Code run
There aren't as many errors as I thought -- 9, after the React fix in DefinitelyTyped. They are all very good errors except one questionable `__metadata` pattern. Overall, the RWC run indicates that this is a good change.

I will look at the errors from DefinitelyTyped next.

## Good

### Misleading names

There are two kinds of misleading names: 
1. Wishful thinking: surely `IHeaderViewOptions` is assignable to `IViewOptions`, right?
2. Namespaces: I bet `Common.Controls.Legacy.ITreeDataItem` is assignable to `Common.Controls.Legacy.Grid.ITreeDataItem` or at least `MyProject.ITreeDataItem`, right?

These are *really* hard to catch without the aid of a compiler. And Typescript wasn't checking assignability at all for weak targets. Now it catches the most egregious errors.

This change found one instance of wishful thinking in VS Code and one instance of the namespace problem in another internal Microsoft project. (Who else would have four layers of legacy controls? (Don't answer that.))

### Confusing types

There are plenty of types which aren't assignable but look like they might be at first glance. For example: `IJSONSchema | IJSONSchema[]` to `IJSONSchema`.

The compiler now catches more of these. It found 2 in VS Code, and 2 in 2 other internal Microsoft projects.

### Misleading `implements` 

See above for discussion of the React `ComponentLifecycle` error.  Two projects this error, which is now fixed in DefinitelyTyped.

### Total Spaceship Guy

There are plenty of incorrect assignments that are obvious if you ever notice them. Things like `string[]` to `IJSONSchema`, or `Promise<IBranch>` to `IBranch` (both from VS Code).

The compiler found 3 of these in VS Code and 1 in another Microsoft project.

## Ugly

Another internal MS project had a nearly-empty type called `Target` with a single optional property named `__metadata`. I guess that `Target` is intended to mean that nearly any type in the project might have a `__metadata` property with something interesting on it. And before this change, you could cast anything to `Target` and check whether `__metadata` actually exists.

## Bad 

There were no bad errors!

# Fixes

* Misleading names and totally wrong assignments are both easy to fix. 
* Confusing types can be easy to fix if you know the codebase, but otherwise might lead to hours of changing types here and there until the project compiles again. 
* Misleading implements are not obvious to fix, but I think that the fix I ended up using for React will always work: remove the `implements` from the class and instead merge it with an interface that extends the previously-implemented interface.
* I think the fix for the common `__metadata` type is similar: lots of classes in the project need to merge with an interface that extends the `__metadata` type in order to be assignable to it.

**Comment by sandersn** [comment]
# Summary of DefinitelyTyped run

Again, there are not as many errors as I thought. Of 3095 projects total, there were 20 errors in definition files and about 80 errors in test files. I haven't looked at the test files yet, but I didn't find any new error patterns in the definitions. Here's the breakdown:

## Good 

### Misleading names

Wishful thinking: kendo-ui (3 times), react-leaflet, react-native

### Misleading `implements`

bookshelf, mongoose, js-yaml, mapbox-gl, polymer

### Totally wrong

react-mdl (10 times).

**Comment by RyanCavanaugh** [comment]
Fantastic. Those are much better results than I expected 👍

**Comment by sandersn** [comment]
Yes, there are few enough errors that I am going to go fix up DefinitelyTyped right now.

Note that the Ugly error pattern is very similar to one in the compiler I had to fix in this PR. `SymbolLinks` is a weak type so you can cast *anything* to it. `getRootSymbols` checks whether a `Symbol` is a `TransientSybmol`, then casts `Symbol` to `SymbolLinks`. But `Symbol` doesn't inherit from `SymbolLinks`. `TransientSymbol` does, so the fix is to cast to `TransientSymbol` instead. It's likely that the other MS project has a similar simple fix.

**Comment by mihailik** [comment]
What is the canonical recommended way of solving React-like false error?

I.E. base class with all methods optional, derived class doesn't implement any of those optional methods.

**Comment by sandersn** [comment]
@mihailik class-interface merging. Remove the implements clause from the class declaration and add a line preceding it:

```ts
interface C extends Optional {}
```

**Comment by sandersn** [comment]
@mihailik If you are writing new code or able to restructure the old code, I think it's better to extend a class instead of merging an extending interface. It should be a lot less confusing for people who haven't seen much merging.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -8977,17 +8978,35 @@ namespace ts {
...
                     const related = isRelatedTo(source, targetType, reportErrors);
                     if (!related) {
+                        disableWeakTypeErrors = saveDisableWeakTypeErrors;
                         return Ternary.False;
                     }
                     result &= related;
                 }
-                return result;
+                disableWeakTypeErrors = saveDisableWeakTypeErrors;
+                return reportAssignmentToWeakIntersection(source, target, reportErrors) ? Ternary.False : result;
```
Comment (yuit) [inline src/compiler/checker.ts]: I may miss something here ->why would we call "reportAssignmentToWeakIntersection" without checking first that `disableWeakTypeErrors ` is false

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -9355,9 +9381,30 @@ namespace ts {
...
+                    return Ternary.False;
+                }
                 return result;
             }
 
+            /**
+             * A type is 'weak' if it is an object type with at least one optional property
+             * and no required properties, call/construct signatures or index signatures
+             */
+            function isWeak(type: Type) {
```
Comment (yuit) [inline src/compiler/checker.ts]: nit: could you rename as `isWeakType`?

Comment (sandersn) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 2 above)
Comment (sandersn) [inline src/compiler/checker.ts]: I assume that you always want to check the combined intersection type. Put another way, I assume that the only reason that `disableWeakTypeErrors` would be false is to prevent checking of individual intersection constituents, which doesn't apply to `reportAssignmentToWeakIntersection`. 

tl;dr It doesn't need to check `disableWeakTypeErrors` because it doesn't apply.

Comment (yuit) [inline src/compiler/checker.ts]: Oic. Jsut to make sure I understand this correctly because `reportAaassignmentToWeakIntersection` does the check in line 8999 and 9000 -> checking that all constituents are weak type ? .....would you mind add this comment in? it wasn't obvious jsut from looking at the caller

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -8701,6 +8701,7 @@ namespace ts {
             let expandingFlags: number;
             let depth = 0;
             let overflow = false;
+            let disableWeakTypeErrors = false;
```
Comment (yuit) [inline src/compiler/checker.ts]: talk offline :: this should be renamed 🌵

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -8977,17 +8978,35 @@ namespace ts {
...
+                disableWeakTypeErrors = saveDisableWeakTypeErrors;
+                return reportAssignmentToWeakIntersection(source, target, reportErrors) ? Ternary.False : result;
+            }
+
+            function reportAssignmentToWeakIntersection(source: Type, target: IntersectionType, reportErrors: boolean) {
+                const needsWeakTypeCheck = source !== globalObjectType && getPropertiesOfType(source).length > 0 && every(target.types, isWeakType);
+                if (!needsWeakTypeCheck) {
+                    return false;
+                }
+                const hasSharedProperty = forEach(
```
Comment (yuit) [inline src/compiler/checker.ts]: nit: please add comment on why you don't use `isrelatedTo`

**Comment by sandersn** [comment]
Addendum to DefinitelyTyped survey: mithril turns out to have an Ugly error that results from this change:

```ts
// @filename: index.d.ts
interface Lifecycle<Attrs, State> {
  onX?(attrs: Attrs): void;
  onY?(state: State): void;
  // etc
}
interface Component<Attrs, State extends Lifecycle<Attrs, State>> {
  view(this: State, vnode: VNode<Attrs, State>): Children | null | void
}
```

((Simplified from the original.))[https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mithril/index.d.ts] Basically, `Component` takes a state object that has some custom state fields, plus may optionally implement some callbacks &mdash; `onX` and `onY` in the above simplified example.

```ts
// @filename: mithril-test.ts
interface State {
  ok: number
  // could also implement 'onX' here
}
const c1: Component<{}, State> = 
  view({attrs, state}) {
    this.ok // 'this' has a member 'ok'
  }
}
```

The problem is that in the common case, `State` doesn't implement any of the callbacks, so fails the weak type check. But it *could* implement some of the callbacks, in which case the type should be checked at some point. The authors could ask that their users instead write `interface State extends Lifecycle<{}, State>`, but this would be a breaking change for their users.

The workaround that I came up with is to add a string index signature to `Lifecycle`, so that it's no longer a weak type. But I'm not sure if this solution generalises, or even appliess well to this case.

### Appendix 

Complicating matters is that the Component can actually be used as State as well:

```ts
const c2 = {
  ok: 12
  view({attrs, start}) {
    this.ok // 'this' should have a member 'ok' (but doesn't right now)
  }
  // could also implement 'onX' here
}
```

But I believe this is only solvable with `ThisType<T>`.

**Comment by sandersn** [comment]
Thanks for the improvements @ahejlsberg. I'll merge as soon as CI finishes.

**Comment by xogeny** [comment]
Given that this breaks existing code and it will take time for libraries to catch up, it would be nice if there were a flag in `tsconfig.json` to disable this checking for now.  I don't see any mention of that here...any chance that was actually done and I'm just not seeing it?

**Comment by RyanCavanaugh** [comment]
@xogeny we generally don't provide flags for cases where we believe all the breaks to be "correct". Are there certain patterns where you need to do non-bugfix work to fix an error?

**Comment by xogeny** [comment]
@RyanCavanaugh the issue is that there are libraries out there whose `.d.ts` files break this rule so it is now impossible to use them (at least with 2.4).  These are libraries I have no control over and it may take time for the authors to get around to addressing this (if at all).  I understand your position, I was just hoping there was some simple way to deal with this in the interim while libraries are catching up.

**Comment by kitsonk** [comment]
> the issue is that there are libraries out there whose `.d.ts`

Can you provide examples?  Because those people have errors in their code/definitions and should really be fixed ASAP.

**Comment by xogeny** [comment]
The one that is currently an issue for me is:

https://github.com/palantir/blueprint/issues/1237

But here is another example:

https://github.com/ceolter/ag-grid/issues/1708 (closed, but without a fix...hmmm)

I've also seen some issues with `@types/react` definitions as well (I don't have a link for that, I've just seen them while attempting to upgrade).

BTW, I'm not saying these aren't errors in their code.  I'm not even saying there should be an option to disable the weak checks.  I was simply asking if such an option exists.  If it doesn't exist, then I'll just wait until blueprint 1.21.0 comes out.

**Comment by sandersn** [comment]
We are working on the definitely typed failures from 2.4. I think I got the last of the weak type failures in #17303. For packages that bundle their own types, you can work around the un-updated libraries by manually casting to the weak type: `myThing as WeakType` because casts to weak types always succeed.

**Comment by xogeny** [comment]
So I thought this would get resolved upstream in `@blueprintjs/core`.  But I see that it has been bumped down the list of milestones quite a ways now (https://github.com/palantir/blueprint/issues/1237, https://github.com/palantir/blueprint/milestone/26).

I don't really want to wait a month for this, so I'm trying to figure out a way to deal with this in the meantime.  What I don't understand about this `myThing as WeakType` workaround is how does it apply to existing libraries.  When I compile against `1.21.0` of `@blueprintjs/core`, I get this error:

```
node_modules/@blueprintjs/core/dist/components/collapse/collapse.d.ts(46,32): error TS2559: Type '{ className: string; s
tyle: { height: string; overflowY: string; transition: string; }; }' has no properties in common with type 'DOMAttribute
s<Element>'.
```

How does casting help me.  This is an error in the typings themselves, not in my code.  I'm guessing that in order to get this to compile, I'm going to have to go in and modify `collapse.d.ts` manually for every install until the Blueprint team fixes this issue and publishes a new version.  Somebody please tell me I'm missing something here...?

**Comment by sandersn** [comment]
This is actually a problem with `React.DOMElement`. It says that its first type argument must be a `DOMAttributes`, but in actual usage, people always pass either `HTMLAttributes` or `SVGAttributes`. Line 46 in collapse.d.ts passes `HTMLAttributes`. The fix is to change the type paramter constraint in `React.DOMElement`. I'll do that and add this test case to `@types/react` so that after it's merged you can just npm install to pull it in.

**Comment by sandersn** [comment]
DefinitelyTyped/DefinitelyTyped#17581 will *almost* fix this. Unfortunately, the improved types expose another bug in `collapse.d.ts`: its `overflowY` is just `string`, but needs to be `"auto" | ...` to match the actual type in `HTMLAttributes`.

This is a much simpler fix, though. I'll update palantir/blueprint#1237.

**Comment by levenleven** [comment]
Shouldn't this also prevent passing primitives as a parameter?

`setup(1); // no error`

I know I can fix this by defining setup as `function setup(options: Options & object) { }`, but maybe it should be a "default"?

**Comment by sandersn** [comment]
@levenleven #16343 should have fixed this, but I don't think it shipped in the RC. Are you running 2.4.1 (the official release) or 2.4.0 (the RC).

**Comment by levenleven** [comment]
@sandersn Ok, thanks. Tried in playground, not sure which version is running there.

**Comment by sandersn** [comment]
Oops, I think the playground may still be running 2.3. @DanielRosenwasser did you update to 2.4 yet? Is it possible to add the version in small print when you do?

**Comment by masaeedu** [comment]
I would like to define a type for "non-thenables". I was assuming the following would work:

```
type NotThenable = { then?: never };

const x: NotThenable = 10
```

However I get the "has no properties in common with" type error. After this change, how do I define a type for all values that do not have a `then` property?

**Comment by RyanCavanaugh** [comment]
@masaeedu 

```ts
type NotThenable = { then?: never; [others: string]: any } | number | string | boolean | symbol;

const x: NotThenable = 10;
const x1: NotThenable = {};
const x2: NotThenable = { m: 10 };
const x3: NotThenable = { then: () => null };
```

**Comment by masaeedu** [comment]
@RyanCavanaugh I was assuming ` | number | string | boolean | symbol | etc.` is equivalent to `{}`. Does this mean `{}` is no longer the top type?

**Comment by RyanCavanaugh** [comment]
I don't understand the question. What is `etc`? Clearly `{}` isn't the same as ` { then?: never; [others: string]: any }`
